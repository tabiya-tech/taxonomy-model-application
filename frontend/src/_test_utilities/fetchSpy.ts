export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
