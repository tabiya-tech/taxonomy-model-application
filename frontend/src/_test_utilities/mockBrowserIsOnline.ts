import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
