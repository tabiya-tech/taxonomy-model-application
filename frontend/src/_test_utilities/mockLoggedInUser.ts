import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
