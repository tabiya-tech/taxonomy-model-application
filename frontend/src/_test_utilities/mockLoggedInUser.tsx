import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
