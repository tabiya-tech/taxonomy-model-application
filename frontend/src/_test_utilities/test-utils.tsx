// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
