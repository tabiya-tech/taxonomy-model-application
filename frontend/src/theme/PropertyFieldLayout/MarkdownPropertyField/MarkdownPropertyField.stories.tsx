import { Meta, StoryObj } from "@storybook/react";
import MarkdownPropertyField from "src/theme/PropertyFieldLayout/MarkdownPropertyField/MarkdownPropertyField";

const meta: Meta<typeof MarkdownPropertyField> = {
  title: "components/MarkdownPropertyField",
  component: MarkdownPropertyField,
  tags: ["autodocs"],
  args: {},
};

export default meta;

type Story = StoryObj<typeof MarkdownPropertyField>;

const headingsMarkdown = `
# Heading Level 1
## Heading Level 2
### Heading Level 3
#### Heading Level 4
##### Heading Level 5
###### Heading Level 6
`;

const textStylesMarkdown = `
- **Bold text**
- *Italic text*
- ~~Strikethrough~~
`;

const listsMarkdown = `
### Unordered List
- Item 1
- Item 2
- Item 3

### Ordered List
1. First item
2. Second item
3. Third item
`;

const linksMarkdown = `
Here is a [link to an example](https://example.com).\n
Here is a urn link [Example URN](urn:esco:occupation:1234).
`;

export const Shown: Story = {
  args: {
    label: "Overview of Markdown ",
    text: `
This example showcases basic Markdown elements like headings, lists, text styles, and links.

## Simple Heading
This is a simple text with a list:
- Item **1**
- Item **2**
- Item **3**

## Text Styles
- **Bold text**
- *Italic text*
- ~~Strikethrough~~

## Links
Here is a link to [Example URL](https://example.com).\n
And here is a link with a URN: [Example URN](urn:esco:occupation:1234).
    `,
  },
};

export const ShownWithLinks: Story = {
  args: {
    label: "Links",
    text: linksMarkdown,
  },
};

export const ShownWithHeadings: Story = {
  args: {
    label: "Headings",
    text: headingsMarkdown,
  },
};

export const ShownWithTextStyles: Story = {
  args: {
    label: "Text Styles",
    text: textStylesMarkdown,
  },
};

export const ShownWithLists: Story = {
  args: {
    label: "Lists",
    text: listsMarkdown,
  },
};

export const ShownCombinedMarkdown: Story = {
  args: {
    label: "Comprehensive Markdown Example",
    text: `
This is an example of many Markdown features combined together. It includes headings, lists, text styles, links, tables, and images.
## Headings
${headingsMarkdown}

## Text Styles
${textStylesMarkdown}

## Lists
${listsMarkdown}

## Links
${linksMarkdown}

## Blockquote
> "This is a blockquote used for emphasizing content."

## Email
You can contact us at [fake@example.com](mailto:fake@example.com).

## Images
![Tabiya logoI](https://tabiya.tech/images/logo.svg)

## Table
| Column 1 | Column 2 |
| -------- | -------- |
| Row 1    | Data 1   |
| Row 2    | Data 2   |
| Row 3    | Data 3  |
    `,
  },
};
