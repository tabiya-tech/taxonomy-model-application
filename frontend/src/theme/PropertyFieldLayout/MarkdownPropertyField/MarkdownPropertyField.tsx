import { Typography } from "@mui/material";
import React from "react";
import ReactMarkdown from "react-markdown";
import remarkGfm from "remark-gfm";
import PropertyFieldLayout from "src/theme/PropertyFieldLayout/PropertyFieldLayout";

export interface MarkdownPropertyFieldProps {
  label?: string;
  text: string;
  "data-testid"?: string;
  fieldId?: string;
}

const uniqueId = "0af822fc-07e3-4032-8276-d7ebf6dc4631";

export const DATA_TEST_ID = {
  MARKDOWN_PROPERTY_FIELD_TEXT: `markdown-property-field-text-${uniqueId}`,
};

const handleLink = (props: any) => {
  return (
    <a {...props} target="_blank" rel="noreferrer noopener">
      {props.children}
    </a>
  );
};

const MarkdownPropertyField = (props: Readonly<MarkdownPropertyFieldProps>) => {
  return (
    <PropertyFieldLayout title={props.label ?? ""} data-testid={props["data-testid"]} fieldId={props?.fieldId ?? ""}>
      <Typography sx={{ "& > *": { margin: "0" } }} data-testid={DATA_TEST_ID.MARKDOWN_PROPERTY_FIELD_TEXT}>
        <ReactMarkdown
          remarkPlugins={[remarkGfm]}
          components={{
            a: handleLink,
          }}
        >
          {props.text}
        </ReactMarkdown>
      </Typography>
    </PropertyFieldLayout>
  );
};

export default MarkdownPropertyField;
