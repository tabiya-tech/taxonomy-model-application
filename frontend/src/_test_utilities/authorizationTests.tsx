import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
