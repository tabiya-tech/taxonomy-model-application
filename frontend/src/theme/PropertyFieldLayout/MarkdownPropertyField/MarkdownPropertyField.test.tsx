// mute the console
import "src/_test_utilities/consoleMock";

import React from "react";
import { render, screen } from "src/_test_utilities/test-utils";
import MarkdownPropertyField, {
  handleLink,
  handleTransform,
} from "src/theme/PropertyFieldLayout/MarkdownPropertyField/MarkdownPropertyField";

// mock the PropertyFieldLayout component
jest.mock("src/theme/PropertyFieldLayout/PropertyFieldLayout", () => {
  return {
    __esModule: true,
    default: jest
      .fn()
      .mockImplementation((props: { title: string; "data-testid": string; children: React.ReactNode }) => {
        return (
          <div data-testid={props["data-testid"]}>
            <span>{props.title}</span>
            <span>{props.children}</span>
          </div>
        );
      }),
  };
});

// mock the ReactMarkdown component
jest.mock("react-markdown", () => {
  return jest.fn().mockImplementation((props) => {
    const { remarkPlugins, urlTransform, components, ...rest } = props;
    return <span {...rest} />;
  });
});

describe("MarkdownPropertyField", () => {
  test("should render with provided markdown", () => {
    /// GIVEN a text with markdown
    const givenText = "# Header\n\nThis is a paragraph with **bold** and *italic* text.\n\n- item 1\n- item 2";
    // AND a label
    const givenLabel = "foo";
    // AND a valid fieldId
    const givenFieldId = "field-id";
    // AND given data-testid
    const givenDataTestId = "markdown-details";

    // WHEN the component is rendered
    render(
      <MarkdownPropertyField label={givenLabel} text={givenText} data-testid={givenDataTestId} fieldId={givenFieldId} />
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the component to be displayed
    const fieldPropertyComponent = screen.getByTestId(givenDataTestId);
    expect(fieldPropertyComponent).toBeInTheDocument();
    // AND the component to match the snapshot
    expect(fieldPropertyComponent).toMatchSnapshot();
  });

  test("should call handleTransform for each link", () => {
    // GIVEN a urn link
    const givenUrl = "urn:esco:occupation:1234";

    // WHEN the component is rendered
    render(<MarkdownPropertyField text={givenUrl} />);

    // THEN expect the handleTransform function to be called
    const result = handleTransform(givenUrl);
    expect(result).toEqual(givenUrl);
  });

  describe("handleLink", () => {
    test("should transform urn:esco links correctly", () => {
      // GIVEN a urn link
      const props = {
        href: "urn:esco:occupation:1234",
        children: "ESCO link",
      };

      // WHEN the handleLink function is called
      render(handleLink(props));

      // THEN expect the link to be rendered correctly
      const linkElement = screen.getByRole("link", { name: /ESCO link/i });
      expect(linkElement).toBeInTheDocument();
      // AND the link to have the correct attributes
      expect(linkElement).toHaveAttribute("href", "https://data.europa.eu/esco/occupation/1234");
      expect(linkElement).toHaveAttribute("target", "_blank");
      // AND the link to have the correct text content
      expect(linkElement).toHaveTextContent("ESCO link");
    });

    test("should render regular links without transformation", () => {
      // GIVEN a normal link
      const props = {
        href: "https://example.com",
        children: "Example link",
      };

      // WHEN the handleLink function is called
      render(handleLink(props));

      // THEN expect the link to be rendered correctly
      const linkElement = screen.getByRole("link", { name: /Example link/i });
      expect(linkElement).toBeInTheDocument();
      // AND the link to have the correct attributes
      expect(linkElement).toHaveAttribute("href", "https://example.com");
      expect(linkElement).toHaveAttribute("target", "_blank");
      // AND the link to have the correct text content
      expect(linkElement).toHaveTextContent("Example link");
    });

    test("should render unknown urn links without transformation", () => {
      // GIVEN an unknown urn link
      const props = {
        href: "urn:unknown:resource:1234",
        children: "Unknown URN link",
      };

      // WHEN the handleLink function is called
      render(handleLink(props));

      // THEN expect the link to be rendered correctly
      const linkElement = screen.getByRole("link", { name: /Unknown URN link/i });
      expect(linkElement).toBeInTheDocument();
      // AND the link to have the correct attributes
      expect(linkElement).toHaveAttribute("href", "urn:unknown:resource:1234");
      expect(linkElement).toHaveAttribute("target", "_blank");
      // AND the link to have the correct text content
      expect(linkElement).toHaveTextContent("Unknown URN link");
    });
  });
});
