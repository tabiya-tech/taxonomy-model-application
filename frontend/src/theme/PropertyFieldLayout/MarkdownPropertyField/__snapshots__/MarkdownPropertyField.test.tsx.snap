// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownPropertyField should render with provided markdown 1`] = `
<div
  data-testid="markdown-details"
>
  <span>
    foo
  </span>
  <span>
    <span
      class="MuiTypography-root MuiTypography-body1 css-swfwp-MuiTypography-root"
      data-testid="markdown-property-field-text-0af822fc-07e3-4032-8276-d7ebf6dc4631"
    >
      <span>
        # Header

This is a paragraph with **bold** and *italic* text.

- item 1
- item 2
      </span>
    </span>
  </span>
</div>
`;
